GCE discovery

gottwald opened the pull request:
This enables discovery for GCE Scraping targets

Simple example scrape config:
```
scrape_configs:
  - job_name: node-exporter
    gce_sd_configs:
      - project: your-project-id
        zone: europe-west1-b
        port: 9100
    relabel_configs:
      - source_labels: ['__meta_gce_instance_name']
        target_label: 'instance_name'
```

Authentication is handled by the google libs DefaultClient and work like this:
(copied from the lib documentation)
```
It looks for credentials in the following places, preferring the first location found:
1. A JSON file whose path is specified by the
   GOOGLE_APPLICATION_CREDENTIALS environment variable.
2. A JSON file in a location known to the gcloud command-line tool.
   On Windows, this is %APPDATA%/gcloud/application_default_credentials.json.
   On other systems, $HOME/.config/gcloud/application_default_credentials.json.
3. On Google App Engine it uses the appengine.AccessToken function.
4. On Google Compute Engine and Google App Engine Managed VMs, it fetches
   credentials from the metadata server.
   (In this final case any provided scopes are ignored.)
```

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/prometheus/prometheus/1769)
<!-- Reviewable:end -->

brian-brazil commented on config/config.go:
```diff
@@ -828,6 +836,47 @@ func (c *KubernetesSDConfig) UnmarshalYAML(unmarshal func(interface{}) error) er
+	// Filter: Can be used optionally to filter the instance list by other criteria.
+	// Syntax of this filter string is described here in the filter query parameter section:
+	// https://cloud.google.com/compute/docs/reference/latest/instances/list
+	Filter string `yaml:"filter,omitempty"`
```
We generally see things like this only as a performance optimisation, is there anything possible with this that won't be possible with relabelling?

gottwald replied:
At the moment most of it can be done by relabelling, except filtering for instance tags.
But the next Google API change (which is currently in beta) will allow to filter for more complex data, like nested fields in their API and also labels (random key value pairs that are assigned to the instances).
Not sure how this could be done with relabelling only.

brian-brazil replied:
You should expose all this information as metadata for relabelling. If it's useful for filtering, it's likely also useful as labels too.

fabxc replied:
If filtering can be enabled user-defined in such a straightforward way, I'm
voting for keeping it. Should still be exposed in labels to a sane degree.

On Tue, Jun 28, 2016, 8:17 PM Brian Brazil <notifications@github.com> wrote:

> In config/config.go
> <https://github.com/prometheus/prometheus/pull/1769#discussion_r68812638>:
>
> > @@ -828,6 +836,47 @@ func (c *KubernetesSDConfig) UnmarshalYAML(unmarshal func(interface{}) error) er
> >  	return nil
> >  }
> >
> > +// GCESDConfig is the configuration for GCE based service discovery.
> > +type GCESDConfig struct {
> > +	// Project: The Google Cloud Project ID
> > +	Project string `yaml:"project"`
> > +
> > +	// Zone: The zone of the scrape targets.
> > +	// If you need to configure multiple zones use multiple gce_sd_configs
> > +	Zone string `yaml:"zone"`
> > +
> > +	// Filter: Can be used optionally to filter the instance list by other criteria.
> > +	// Syntax of this filter string is described here in the filter query parameter section:
> > +	// https://cloud.google.com/compute/docs/reference/latest/instances/list
> > +	Filter string `yaml:"filter,omitempty"`
>
> You should expose all this information as metadata for relabelling. If
> it's useful for filtering, it's likely also useful as labels too.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/prometheus/prometheus/pull/1769/files/aaca980fcffb0da8ac9d3711ab33e4abfb536610..aaf0f438d9011e3d02fff333c738551ab9d491d0#r68812638>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AEuA8q16v37qvb_mNNiIs-uzq7HLVOVZks5qQWVQgaJpZM4JAGOO>
> .
>

gottwald replied:
How would you like me to expose the tags? Since there can be several tags on one instance and there is no real key-value pair (the tag is the value), how should I expose them?

Should they be enumerated like:
```
__meta_gce_tag_0: my-tag
__meta_gce_tag_1: my-other-tag
```
Of course this would need to be sorted, because we can't be sure that the api answer has always the same order of tags.
This way could also make relabelling harder, especially if tags change.

Or we could use the value twice like:
```
__meta_gce_tag_my-tag: my-tag
__meta_gce_tag_my-other-tag: my-other-tag
```

Any other ideas or suggestions how to do this in a reasonable way?

brian-brazil replied:
`,tag1,tag2,tag3,` as a value is our standard.

gottwald replied:
So just a joined string. Just to be sure, "," at the beginning and at the end?

grobie replied:
Yes, that makes regular expressions easier. See
http://www.robustperception.io/little-things-matter/

On Thu, Jun 30, 2016 at 2:31 PM, Ingo Gottwald <notifications@github.com>
wrote:

> In config/config.go
> <https://github.com/prometheus/prometheus/pull/1769#discussion_r69185745>:
>
> > @@ -828,6 +836,47 @@ func (c *KubernetesSDConfig) UnmarshalYAML(unmarshal func(interface{}) error) er
> >  	return nil
> >  }
> >
> > +// GCESDConfig is the configuration for GCE based service discovery.
> > +type GCESDConfig struct {
> > +	// Project: The Google Cloud Project ID
> > +	Project string `yaml:"project"`
> > +
> > +	// Zone: The zone of the scrape targets.
> > +	// If you need to configure multiple zones use multiple gce_sd_configs
> > +	Zone string `yaml:"zone"`
> > +
> > +	// Filter: Can be used optionally to filter the instance list by other criteria.
> > +	// Syntax of this filter string is described here in the filter query parameter section:
> > +	// https://cloud.google.com/compute/docs/reference/latest/instances/list
> > +	Filter string `yaml:"filter,omitempty"`
>
> So just a joined string. Just to be sure, "," at the beginning and at the
> end?
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/prometheus/prometheus/pull/1769/files/aaca980fcffb0da8ac9d3711ab33e4abfb536610..aaf0f438d9011e3d02fff333c738551ab9d491d0#r69185745>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AAANaJIJ3K4TQiHKRbhcURmCKY-LGiBMks5qRAtsgaJpZM4JAGOO>
> .
>

gottwald commented:
Added tags meta labels like the consul discovery does.

gottwald commented:
Any clues how to fix the circle CI build?
Looks like there are some missing header files for the freebsd crossbuild.
The only package I found with the header files included isn't available for the amd64 arch.

grobie commented:
@gottwald Does this change introduce a new dependency on CGO?

gottwald commented:
@grobie I don't think so. Did a fgrep over the vendored sources and only found these uses of cgo:
```
src/github.com/prometheus/prometheus/vendor/golang.org/x/sys/unix/types_darwin.go:import "C"
src/github.com/prometheus/prometheus/vendor/golang.org/x/sys/unix/types_dragonfly.go:import "C"
src/github.com/prometheus/prometheus/vendor/golang.org/x/sys/unix/types_freebsd.go:import "C"
src/github.com/prometheus/prometheus/vendor/golang.org/x/sys/unix/types_linux.go:import "C"
src/github.com/prometheus/prometheus/vendor/golang.org/x/sys/unix/types_netbsd.go:import "C"
src/github.com/prometheus/prometheus/vendor/golang.org/x/sys/unix/types_openbsd.go:import "C"
src/github.com/prometheus/prometheus/vendor/golang.org/x/sys/unix/types_solaris.go:import "C"
```

But `golang.org/x/sys/unix` was already vendored before this change.
And also I also can't find any use of that package in the code and deps I added.
Looks like it's only used by `fsnotify`, which is used by the file based discovery.

gottwald commented:
@grobie Any idea how we can proceed on this?
I rebuilt the golang-builder with all freebsd header files in it and modified CPPFLAGS for freebsd, with the result that the header files are found but I get C compile errors. I'm also not really sure that I'm doing the right thing here since I've no clue about cross compiling for freebsd with cgo enabled.

grobie commented:
Maybe @sdurrheimer has a hunch of what the problem might be here.

I'll need to schedule some time to debug this. This won't happen before next week (due to PromCon and me flying to Berlin).

gottwald commented:
Ok, some more info on this:
Now I'm 100% sure this is a cross compile issue.
Just tested on 2 vagrant boxes (freebsd 10.3 and 11).
After some Makefile fixes due to incompatibilities, I'm able to do 2 successful builds with this branch and cgo enabled.

bluepeppers commented:
Might be of interest to anyone browsing this ticket, using the file discoverer: https://github.com/QubitProducts/prometheus_gce_sd (not to detract in any way from the work here, which would make life much easier in many ways 👍 )

sdurrheimer commented:
Having a dependency is a thing, using its code another.

I already encountered this error when cross-compiling others of our golang components. I also tried to fix it by adding those header files, but then there is even more errors.

It would be nice to detect the part that needs those headers, otherwise I really don't know what we can do except removing freebsd (and probably openbsd and dragonfly) from the list of crossbuilt platforms for Prometheus.

brian-brazil commented:
I don't think we can drop FreeBSD. Have we filed a bug upstream with GCE?

gottwald commented:
@brian-brazil I don't quite see how this is a upstream bug when prometheus and the gce lib compiles perfectly fine on a freebsd system but fails on this cross compilation stack.
I'd rather say somethings missing/not right with the cross compiler toolchain.

sdurrheimer commented on vendor/vendor.json:
```diff
@@ -3,6 +3,18 @@
 	"package": [
 		{
+			"path": "appengine",
+			"revision": ""
```
No revision for those ?

gottwald replied:
I'll look at that. I wonder why govendor didn't put one in, maybe I was holding it wrong...

gottwald replied:
Those were internal deps for appengine.
All those without revision were guarded by an appengine only build tag.
I added appengine to the ignore tags of govendor which fixed those ones.
There was only one left which was external and accidentally not moved to the vendor folder. Fixed that as well.

sdurrheimer commented:
I'm investigating the problem.

Also you will need to rebase your work against master.

gottwald commented:
@sdurrheimer rebased against master

gottwald commented:
@sdurrheimer I was trying to build a freebsd crossbuild toolchain on top of the current golang-builder.
Was following this article: https://linux-tips.org/t/how-to-build-a-freebsd-toolchain-for-linux/288 with somewhat newer versions, but I always failed during gcc compiling due to header file mess.
I could share a Dockerfile if you care.
The interesting part is, that according to this article if you manage to get gcc built that way you should end up with a `x86_64-pc-freebsd10-gcc` binary which we're missing atm.

sdurrheimer commented:
@gottwald Interesting, I will take a look. Thanks.

The thing to keep in mind, if we encounter this error with freebsd, we will also have it with openbsd and dragonfly.

sdurrheimer commented:
Ok, I think I need to improve the `golang-builder` in order to not always enable CGO, and enable it when a project needs it, probably with a `promu` flag.

gottwald commented:
Cross-build without CGO would make thinks much easier.
Does prometheus actually need cgo?
Just found a few `import C` lines in all the `golang.org/x/sys/unix/types_*` files used in the dependencies.

sdurrheimer commented:
When I started working on the `golang-builder`   `promu` system for our releases, some of our golang components were using CGO, so it was easier to always enable it (although I was petty sure it was mandatory to crossbuild some platforms). Maybe I was wrong.

sdurrheimer commented:
Sorry for the wait, I am still working on the subject to enable `promu / golang-builder` to support both CGO and non-CGO builds.

I just need to make additional tests on our other golang components in order to avoid creating regressions.

sdurrheimer commented:
@gottwald Here we go, I fixed the build system. Could you rebase your work one more time ?

grobie commented:
whoop whoop. Thank you @sdurrheimer! The build looks fine. Any last comments @brian-brazil @fabxc?

gottwald commented:
@sdurrheimer thx for fixing the build system! All green now.

grobie commented on retrieval/discovery/gce.go:
```diff
@@ -0,0 +1,171 @@
+
+func (gd *GCEDiscovery) refresh() (*config.TargetGroup, error) {
+	if gd.isvc == nil {
+		return nil, errors.New("could not refresh gce targets because of previous gce communication setup error")
```
Can you explain when we'd reach this state here?

gottwald replied:
Could happen when the GCE Application Default Credentials are used but the VM is missing the correct auth scope to read from the API for example.

grobie replied:
Still not really clear how we'd reach this function in that case. Isn't there a guard in `Run()` before `refresh()` gets called the first time?

gottwald replied:
That's true, but I'd rather remove the one in `Run()` than this one. Because otherwise `refresh()` has the potential to `panic` because of a nil pointer dereference when called without the guard beforehand. Might bite someone who's refactoring this in the future.

grobie replied:
SGTM

grobie replied:
Or you return an error in the constructor and don't register the discovery at all if the setup fails. It would never be valid to use a discovery type without `isvc` set then. The kubernetes discovery does something similar.

grobie commented:
:+1: We should have tests for all our discovery implementations, but that's not just a problem of the GCE dicsovery.

gottwald commented:
Problem with tests for all cloud vendors is that you can't include it into the normal unit tests, because you need an account with billing enabled for most cloud providers to test.
So this would have to be separated into a end2end test or integration test.
Not sure if any of them offer free tiers for open source projects.

fabxc commented:
You can always fake the API responses of the provider. That's not ideal
coverage, but way better than nothing.

On Fri, Sep 9, 2016 at 8:29 PM Ingo Gottwald <notifications@github.com>
wrote:

> Problem with tests for all cloud vendors is that you can't include it into
> the normal unit tests, because you need an account with billing enabled for
> most cloud providers to test.
> So this would have to be separated into a end2end test or integration test.
> Not sure if any of them offer free tiers for open source projects.
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/prometheus/prometheus/pull/1769#issuecomment-245998958>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AEuA8vKcjJYTScyMp8ZmidDBBNlCHUY3ks5qoaWQgaJpZM4JAGOO>
> .
>

gottwald commented:
@fabxc in case of the GCE API we would have to fake tons of stuff to a degree that the actual testing result has hardly any value because you never test the real thing.
Also things like `basePath` in the client api, that we would need to use in order to inject the url of a fake api handler, are sadly non-exported fields.

juliusv commented:
Yeah, best would be for cloud providers to supply local test versions of their APIs that can be used in integration tests. They partially do that for things like data stores (like https://cloud.google.com/datastore/docs/tools/datastore-emulator), but I don't think it exists for this general GCE functionality.

grobie commented:
Another approach is to record HTTP requests and replay them in further tests. It's essentially a cache to make testing without internet/cluster possible and the option to just throw everything away and validate against the live system from time to time. There are various vcr projects available for golang.

Given we don't have anything in place for any of our discovery mechanisms, I don't think we need to block the gce discovery. We already have an open issue for integration tests https://github.com/prometheus/prometheus/issues/1106

brancz commented:
FWIW I think recording HTTP is a good solution, but best would be if providers would provide contracts, that we could develop against, but it seems like that only happens in a perfect world. But generally I like the approach by [pact](https://github.com/pact-foundation/pact-go).

gottwald commented:
@grobie so will this be merged now?

grobie commented on retrieval/discovery/gce.go:
```diff
@@ -0,0 +1,171 @@
+func (gd *GCEDiscovery) Run(ctx context.Context, ch chan<- []*config.TargetGroup) {
+	defer close(ch)
+
+	ticker := time.NewTicker(gd.interval)
```
Any reason to not move the ticker initialization down above the for loop? No need to start one if we error out early.

gottwald replied:
Absolutely right, will change that.

grobie commented:
One last thought, could you instrument the discovery? An error counter and latency summary would be great.

gottwald commented:
@grobie not sure what you mean by instrument. Do you have an example somewhere to look at?

sdurrheimer commented:
@gottwald Instrument means to monitor the discovery service itself by adding metrics like for example dns lookups and lookups failures for the DNS-SD discovery:

- https://github.com/prometheus/prometheus/blob/master/retrieval/discovery/dns/dns.go#L41-L59
- https://github.com/prometheus/prometheus/blob/737ae60ceaebc4264358e28e377eda061861117e/retrieval/discovery/dns/dns.go#L129

gottwald commented:
@grobie @sdurrheimer added some metrics WDYT?

gottwald commented:
Appended just a little change. Moved from a gauge metric type to a summary for the duration

grobie commented on retrieval/discovery/gce.go:
```diff
@@ -0,0 +1,198 @@
+		case <-ticker.C:
+			t0 := time.Now()
+			tg, err := gd.refresh()
+			gceSDScrapeDuration.Observe(time.Since(t0).Seconds())
```
You could remove the duplicated code by moving the observation into `refresh()` (you'll need named return parameters):

```go
func (gd *GCEDiscovery) refresh() (tg *config.TargetGroup, err error) {
	t0 := time.Now()
	defer func() {
		gceSDScrapeDuration.Observe(time.Since(t0).Seconds())
		gceSDScrapesCount.Inc()
		if err != nil {
			gceSDScrapeFailuresCount.Inc()
		}
	}()

	tg.Source = fmt.Sprintf("GCE_%s_%s", gd.project, gd.zone)
	// ...
}
```

gottwald commented:
- Deduplicated the code.
- Rebased on master
- Resolved the vendor.json conflicts 
- Squashed commits

The single vendor.json file from govendor is a pain for merging and rebasing....

LGTY now?

grobie commented:
Perfect! Thanks for sticking with me ;)
